nixos/modules/virtualisation/hcloud-config.nix

```diff
@@ -0,0 +1,131 @@
+        '';
+      };
+    };
+  };
+
+  config = {
+
+    boot.growPartition = true;
+
+    fileSystems."/" = {
+      device = "/dev/disk/by-label/nixos";
+      fsType = "ext4";
+      autoResize = true;
+    };
+
+    fileSystems."/boot" = lib.mkIf cfg.efi {
+      device = "/dev/disk/by-label/ESP";
+      fsType = "vfat";
+      options = [ "umask=0077" ];
+    };
+
+    boot.loader = {
+      timeout = 1;
+      grub.enable = !cfg.efi;
+      grub.device = "/dev/sda";
+      systemd-boot.enable = cfg.efi;
+      efi.canTouchEfiVariables = true;
+    };
+
+    # Set the hostname from server metadata by default.
+    networking.hostName = lib.mkDefault "";
+
+    # Allow root logins only using the SSH key that the user specified
+    # at server creation time.
+    services.openssh = {
+      enable = true;
+      settings.PermitRootLogin = "prohibit-password";
+    };
+
+    # Per: https://community.hetzner.com/tutorials/install-and-configure-ntp
+    networking.timeServers = [
+      "ntp1.hetzner.de"
+      "ntp2.hetzner.com"
+      "ntp3.hetzner.net"
+    ];
+
+    # Generate network configuration for systemd-networkd.
+    networking.useNetworkd = true;
+    networking.useDHCP = false;
+    systemd.services.systemd-network-generator-hcloud = {
+      wantedBy = [ "systemd-networkd.service" ];
+      # Per: https://github.com/stephank/systemd-network-generator-hcloud/blob/main/systemd-network-generator-hcloud.service.in
+      conflicts = [
+        "shutdown.target"
+        "initrd-switch-root.target"
+      ];
+      wants = [ "network-pre.target" ];
+      before = [
+        "network-pre.target"
+        "systemd-udevd.service"
+        "shutdown.target"
+        "initrd-switch-root.target"
+      ];
+      unitConfig.DefaultDependencies = false;
+      serviceConfig = {
+        ExecStart =
+          "${pkgs.systemd-network-generator-hcloud}/bin/systemd-network-generator-hcloud"
+          # For the below postStart script.
+          + lib.optionalString dynamicHostname " --write-hostname /run/hcloud-hostname"
+          + " --write-public-keys /run/hcloud-public-keys"
+          # For general use.
+          + lib.optionalString cfg.fetchMetadata " --write-metadata /run/hcloud-metadata --write-userdata /run/hcloud-userdata";
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      # Amend the service with a script that applies other metadata properties.
+      postStart = ''
+        ${lib.optionalString dynamicHostname ''
+          if [[ -s /run/hcloud-hostname ]]; then
+            echo "setting hostname..."
+            ${pkgs.nettools}/bin/hostname $(</run/hcloud-hostname)
+            rm /run/hcloud-hostname
+          fi
+        ''}
+
+        if [[ -e /run/hcloud-public-keys ]]; then
+          echo "configuring root ssh authorized keys..."
+          install -o root -g root -m 0700 -d /root/.ssh/
+          mv /run/hcloud-public-keys /root/.ssh/authorized_keys
+        fi
+      '';
```

Comment:
```suggestion
      postStart = lib.optionalString dynamicHostname ''
        if [[ -s /run/hcloud-hostname ]]; then
          echo "setting hostname..."
          ${pkgs.nettools}/bin/hostname $(</run/hcloud-hostname)
          rm /run/hcloud-hostname
        fi
      '' + ''
        if [[ -e /run/hcloud-public-keys ]]; then
          echo "configuring root ssh authorized keys..."
          install -o root -g root -m 0700 -d /root/.ssh/
          mv /run/hcloud-public-keys /root/.ssh/authorized_keys
        fi
      '';
```

Reply:
Note that nixfmt styled this slightly different.